File: tests/fixtures/python/simple.py
import os
import sys
from pathlib import Path
from typing import List, Optional
from . import utils
from ..config import settings

def main() -> None:
    paths: List[Path] = []
    for root, dirs, files in os.walk('.'):
        for f in files:
            paths.append(Path(root) / f)
    print(paths)


def load_config(path: Optional[str] = None) -> dict:
    config_path = path or settings.DEFAULT_CONFIG
    return utils.load_json(config_path)


class FileScanner:
    def __init__(self, root: str) -> None:
        self.root = root

    def scan(self) -> List[str]:
        result = []
        for entry in os.listdir(self.root):
            result.append(entry)
        return result

    def filter(self, ext: str) -> List[str]:
        return [f for f in self.scan() if f.endswith(ext)]


if __name__ == '__main__':
    main()

File: tests/fixtures/rust/simple.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use serde::{Deserialize, Serialize};
use crate::scanner::FileAst;
use crate::graph::types::Node;

mod utils;
mod config;

pub struct FileScanner {
    root: PathBuf,
    extensions: Vec<String>,
}

impl FileScanner {
    pub fn new(root: &Path) -> Self {
        Self {
            root: root.to_path_buf(),
            extensions: vec!["rs".to_string(), "ts".to_string()],
        }
    }

    pub fn scan(&self) -> Vec<PathBuf> {
        let mut result = Vec::new();
        self.walk(&self.root, &mut result);
        result
    }

    fn walk(&self, dir: &Path, out: &mut Vec<PathBuf>) {
        if let Ok(entries) = std::fs::read_dir(dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.is_dir() {
                    self.walk(&path, out);
                } else if let Some(ext) = path.extension() {
                    if self.extensions.contains(&ext.to_string_lossy().to_string()) {
                        out.push(path);
                    }
                }
            }
        }
    }
}

pub fn parse_file(path: &Path) -> Option<FileAst> {
    let content = std::fs::read_to_string(path).ok()?;
    Some(FileAst {
        path: path.to_string_lossy().to_string(),
        language: "rust".to_string(),
        imports: vec![],
        exports: vec![],
        functions: vec![],
        classes: vec![],
    })
}

pub fn index(paths: &[PathBuf]) -> HashMap<String, Node> {
    let mut map = HashMap::new();
    for p in paths {
        if let Some(_ast) = parse_file(p) {
            // placeholder
        }
    }
    map
}
